reana_commons/snakemake.py

```diff
@@ -1,32 +1,64 @@
 # -*- coding: utf-8 -*-
 #
 # This file is part of REANA.
-# Copyright (C) 2021, 2022 CERN.
+# Copyright (C) 2021, 2022, 2024 CERN.
 #
 # REANA is free software; you can redistribute it and/or modify it
 # under the terms of the MIT License; see LICENSE file for more details.
 
 """REANA Snakemake Workflow utils."""
 
 import os
+import sys
 from itertools import filterfalse, chain
 from typing import Any, Dict, List, Optional
-
-from snakemake import snakemake
-from snakemake.dag import DAG
-from snakemake.io import load_configfile
-from snakemake.jobs import Job
-from snakemake.persistence import Persistence
-from snakemake.rules import Rule
-from snakemake.workflow import Workflow
+from pathlib import Path
+
+if sys.version_info >= (3, 11):
+    from snakemake.api import SnakemakeApi
+    from snakemake.settings.types import (
+        ResourceSettings,
+        WorkflowSettings,
+        ConfigSettings,
+        OutputSettings,
+        StorageSettings,
+        DeploymentSettings,
+    )
+else:
+    from snakemake import snakemake
+    from snakemake.dag import DAG
+    from snakemake.io import load_configfile
+    from snakemake.jobs import Job
+    from snakemake.persistence import Persistence
+    from snakemake.rules import Rule
+    from snakemake.workflow import Workflow
 
 from reana_commons.errors import REANAValidationError
+from reana_commons.config import SNAKEMAKE_MAX_PARALLEL_JOBS
 
 
 def snakemake_validate(
     workflow_file: str, configfiles: List[str], workdir: Optional[str] = None
 ):
-    """Validate Snakemake workflow specification.
+    """Validate snakemake workflow."""
+    if sys.version_info >= (3, 11):
+        snakemake_validate_v8(workflow_file, configfiles, workdir)
+    else:
+        snakemake_validate_v7(workflow_file, configfiles, workdir)
+
+
+def snakemake_load(workflow_file: str, **kwargs: Any):
+    """Load snakemake specification."""
+    if sys.version_info >= (3, 11):
+        return snakemake_load_v8(workflow_file, **kwargs)
+    else:
+        return snakemake_load_v7(workflow_file, **kwargs)
+
+
+def snakemake_validate_v7(
+    workflow_file: str, configfiles: List[str], workdir: Optional[str] = None
+):
+    """Old snakemake validate version for python versions < 3.11, it is needed since snakemake8 dropped support for python < 3.11.
```

Comment:
Snakemake 7 workflow validation function, necessary for Python versions < 3.11.

_(to make the docstring shorter)_